File: Data Structures & Algorithms/products-of-array-discluding-self/submission-8.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        prefix = [1] * len(nums)

        for i in range(len(nums)):
            if i > 0:
                prefix[i] = prefix[i - 1] * nums[i - 1]

        curr = 1
        for i in range(len(nums) - 1, -1, -1):
            prefix[i] *= curr
            curr *= nums[i]

        return prefix

File: Data Structures & Algorithms/subsets/submission-8.ts
class Solution {
    /**
     * @param {number[]} nums
     * @return {number[][]}
     */
    subsets(nums: number[]): number[][] {
        const res: number[][] = []

        function dfs(i: any, curr: number[]) {
            if (i >= nums.length) {
                res.push([...curr])
                return
            }

            curr.push(nums[i])
            dfs(i + 1, curr)

            curr.pop()
            dfs(i + 1, curr)
        }
        dfs(0, [])
        return res
    }
}
